File: stack/impl_with_array/stack.java
public class stack {
    int [] arr;
    int top;
    stack(int size) {
        arr = new int[size];
        top = -1;
    }
    // 2 = O(1)
    boolean isEmpty() {
       if(top == -1) // 1
            return true; // 1
       return false; //1 
    }
    //O(1)
    boolean isFull() {
        if(arr.length-1 == top) // 2
            return true; // 1
        return false; // 1
    }
    //O(1)
    boolean push(int data) { 
        if(isFull()) { // O(1)
            System.out.println("stack is full. can't push "+data);
            return false;
        }
        top++;// top = 5
        arr[top] = data; // 0-4
        return true;
    }
    //O(1)
    int pop() {
        if(isEmpty()) {
            System.out.println("empty stack. can't pop element");
            System.exit(-1);
        }
        int item = arr[top]; // arr[-1]
        top--;
        return item;
    }
    //O(1)
    int top() {
        if(isEmpty()) {
            System.out.println("empty stack. can't find top element");
            System.exit(-1);
        }
        int item = arr[top];
        //top--;
        return item;
    }
}

File: stack/impl_with_array/stack.cpp
// Online C++ compiler to run C++ program online
#include <iostream>
#include <climits>
const int max_size = 1024;
using namespace std;
class stack {
    int size;
    int top;
    int data[max_size];
  public :
  stack(int s) {
      size = s;
      top = 0;
  }
  bool isEmpty() {
      if(top == 0)
        return true;
    else
        return false;
  }
  bool isFull() {
    if(top == size)
        return true;
    else
        return false;
  }
  bool push(int d) {
      if(isFull() == true) {
          cout<<"overflow"<<endl;
        return false;
      }
      data[top] = d;
      top++;
      return true;
  }
  int pop() {
      if(isEmpty() == true){
          cout<<"underflow"<<endl;
          return INT_MAX;
      } else {
          top--;
          return data[top];
      }
  }
  void print() {
      for(int i = top-1; i >=0 ; i--) {
          cout<<data[i]<<"\t";
      }
      cout<<endl;
      
  }
  
};
int main() {
    stack s(5); // stack of size 5
    s.push(10);
    s.push(20);
    s.push(30);
    s.push(40);
    s.push(50);
    s.print(); 
    s.push(69); 
    cout<<s.pop()<<endl;
    cout<<s.pop()<<endl;
    s.push(89);
    s.print(); 
   
    return 0;
}
